### PR TITLE
feat(cca): BBR

### LINKED ISSUE #693: BBR congestion control
Google uses it in production, and it should massively increase throughput on bad networks, while simultaneously fighting bufferbloat.

### PR DESCRIPTION
**Pull request by FrankSpitulski**
implements BBR congestion control based on google's quiche. There are some differences in implementation due to chromium's CCA having a different API, especially around sending vs acking packets.

It's a pretty rough first implementation but from my testing it seems to work well. It is very resistant to packet loss especially in high BDP situations.

Please critique and suggest changes as needed.

I've added additional fields and traits for debugging using stats which can be removed if needed.

Closes #693

### CODE REVIEW

File: quinn-proto/src/congestion.rs
Comment (Matthias247) [inline quinn-proto/src/congestion.rs:18]: typedefs are not that useful since they still allow wrong types being used. Newtypes (like `struct PacketNumber(u64)`) are more useful. But I think if such a change is applied, it should be done in the whole crate.

File: quinn-proto/src/congestion.rs
Comment (Matthias247) [inline quinn-proto/src/congestion.rs:30]: It's a trait. No need to add `_` for unused parameters

File: quinn-proto/src/congestion.rs
Comment (Matthias247) [inline quinn-proto/src/congestion.rs:62]: here `_packet_number` is not a `PacketNumber`

File: quinn-proto/src/congestion.rs
Comment (Matthias247) [inline quinn-proto/src/congestion.rs:60]: from an interface point of view the difference between `on_congestion_event` and `on_loss` is not super clear, since a congestion event can be loss too.

File: quinn-proto/src/congestion/bbr.rs
Comment (Matthias247) [inline quinn-proto/src/congestion/bbr.rs:15]: Those should be using Rust style, like `const DEFAULT_MINIMUM_CONGESTION_WINDOW`

File: quinn-proto/src/congestion/bbr.rs
Comment (Matthias247) [inline quinn-proto/src/congestion/bbr.rs:66]: Since this a "simple enum", `self != RecoveryState::NOT_IN_RECOVERY` will be fine. And clippy will probably tell you that the method should take `self` by value since its a small copyable value

File: quinn-proto/src/congestion/bbr.rs
Comment (Matthias247) [inline quinn-proto/src/congestion/bbr.rs:72]: no point in prefixing them with `bbr_` since everything in this file is about bbr?

File: quinn-proto/src/congestion/bbr.rs
Comment (Matthias247) [inline quinn-proto/src/congestion/bbr.rs:88]: you probably want `now.saturating_duration_since(...)` for safety reasons and conciseness

File: quinn-proto/src/congestion/bbr.rs
Comment (Matthias247) [inline quinn-proto/src/congestion/bbr.rs:87]: if the time between those calls is small, this might often end up being zero. Is this expected?

File: quinn-proto/src/congestion/bbr_min_max.rs
Comment (Matthias247) [inline quinn-proto/src/congestion/bbr_min_max.rs:64]: would need an `#[allow(dead_code)]`.  But keeping unused code around is usually an antipattern

File: quinn/src/connection.rs
Comment (Matthias247) [inline quinn/src/connection.rs:462]: `congestion_controller`?  `cca` seems  only obvious for people who work on it. And what's the reason for making this one public - with an expensive copy instead of just returning a reference (which is problematic due to the lock)?

**Review by Matthias247 — COMMENTED** [review]
Very cool to see BBR is integrateable and to hear it performs good! Definitely interested to see this landing.

I did a short skim on this, but comments are mostly stylistic since I don't have any background on BBR.

Since this is rather complex, I would be interested to see some tests as part of these changes. Ideally also in the form that show the benefits compared ot the other congestion controllers.

File: quinn-proto/src/congestion.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion.rs:18]: is that something you'd be interested in? I found it very helpful while developing do know which was which even if it was just in the function def. I had confused the two a few times when developing before I added this.

File: quinn-proto/src/congestion.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion.rs:30]: my bad, cargo fix was complaining so I just let it do its thing.

File: quinn-proto/src/congestion.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion.rs:60]: that's right, a congestion event is both while a loss is just loss

File: quinn-proto/src/congestion.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion.rs]: fixed

File: quinn-proto/src/congestion/bbr.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion/bbr.rs:72]: will remove- I just kept those from the C impl

File: quinn-proto/src/congestion/bbr.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion/bbr.rs:87]: yes, it's just trying to find the max bandwidth overall so it can swallow 0s no issue. I'd welcome more scrutiny on this of course but that's what the google quiche impl seems to do.

File: quinn-proto/src/congestion/bbr_min_max.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion/bbr_min_max.rs:64]: you'd prefer to remove it then? I'm not attached.

File: quinn/src/connection.rs
Comment (FrankSpitulski) [reply quinn/src/connection.rs:462]: this whole section was just for my debugging ease. any changes at all here are fine with me. my only goal was to be able to print out the internal variables once every few seconds while running a transfer.

File: quinn-proto/src/congestion.rs
Comment (djc) [reply quinn-proto/src/congestion.rs:18]: Yeah, I think that would be good to have (but it should likely be a separate PR).

File: quinn-proto/src/congestion/bbr.rs
Comment (djc) [reply quinn-proto/src/congestion/bbr.rs:66]: If we can avoid the need for deriving `PartialEq` (if this is the only use), I'd just as soon use `matches!()` here.

File: quinn-proto/src/congestion.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion.rs:18]: okay, I'll remove it from this one

File: quinn-proto/src/congestion.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion.rs:30]: it's still not happy about it. I think it's because of the empty default impl. I'll add an #[allow(unused_variables)].

File: quinn-proto/src/congestion.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion.rs]: fixed

File: quinn-proto/src/congestion/bbr.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion/bbr.rs]: fixed

File: quinn-proto/src/congestion/bbr.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion/bbr.rs]: fixed, but it doesn't look any more concise to me. can you make sure it's what you intended?

File: quinn-proto/src/congestion/bbr_min_max.rs
Comment (Ralith) [reply quinn-proto/src/congestion/bbr_min_max.rs:68]: I'd prefer to drop it, since otherwise it's at high risk of bitrot. Maybe leave a comment on `update_max` that an inverse version can be defined, and cite a reference for it.

File: quinn-proto/src/congestion/cubic.rs
Comment (Ralith) [inline quinn-proto/src/congestion/cubic.rs:184]: nit: has the false positive been reported to clippy?

File: quinn-proto/src/congestion/cubic.rs
Comment (Ralith) [inline quinn-proto/src/congestion/cubic.rs:167]:
```suggestion
    #[allow(clippy::branches_sharing_code)] // False positive, see discussion inline
```

File: quinn-proto/src/connection/mod.rs
Comment (Ralith) [inline quinn-proto/src/connection/mod.rs:135]: This shouldn't be public. What does "exposed to stats" mean? I don't see any interfaces that would consume this.

File: quinn-proto/src/congestion.rs
Comment (Ralith) [inline quinn-proto/src/congestion.rs:19]: Would this still be in line with the intended semantics?
```suggestion
    /// One or more packets were just sent
```

File: quinn-proto/src/congestion/bbr.rs
Comment (Ralith) [inline quinn-proto/src/congestion/bbr.rs:577]: I'm surprised BBR completely ignores ECN. You're certain that's intended?

File: quinn-proto/src/congestion.rs
Comment (Ralith) [inline quinn-proto/src/congestion.rs:60]: Would it be simpler to add a `lost_bytes: u64` argument to `on_congestion_event`?

File: quinn-proto/src/congestion.rs
Comment (Ralith) [inline quinn-proto/src/congestion.rs:64]: Could this be folded into `on_sent`?

File: quinn-proto/src/congestion/bbr.rs
Comment (Ralith) [inline quinn-proto/src/congestion/bbr.rs:101]: Does this really need to be `pub`?

File: quinn-proto/src/congestion/bbr.rs
Comment (Ralith) [inline quinn-proto/src/congestion/bbr.rs:420]: Does this really need to be `pub`?

File: quinn-proto/src/congestion/bbr.rs
Comment (Ralith) [inline quinn-proto/src/congestion/bbr.rs:603]:
```suggestion
/// Configuration for the [`BBR`] congestion controller
```

File: quinn/src/connection.rs
Comment (Ralith) [inline quinn/src/connection.rs:463]:
```suggestion
    pub fn congestion_state(&self) -> Box<dyn Controller> {
```
How can a user downcast this? Do we really want to expose it, or would it be better off as something you patch in locally when needed for development?

File: quinn-proto/src/congestion/bbr.rs
Comment (Ralith) [inline quinn-proto/src/congestion/bbr.rs:434]: Let's cite the code this is based on, as well as the current revisions of the papers documenting the algorithm. A brief summary of how it compares to Cubic (e.g. "aims for reduced buffer bloat and improved performance over high bandwidth-delay product networks", if that's accurate) would also be nice.

I'm also not sure any of us are currently able to review the *correctness* of this implementation, and afaik its development is still ongoing upstream, so maybe we should also clearly (i.e. in the first line) designate it as experimental.

**Review by Ralith — COMMENTED** [review]
Thanks, this looks pretty cool! Reviewed mostly for style and API impact for now, since I don't presently have time to dig deep into correctness, and it's not clear what the best reference would be to derive such a review from either.

File: quinn-proto/src/congestion/cubic.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion/cubic.rs:184]: it has not. can I get a second opinion on whether clippy is right here? I'd rather not cite clippy as having an error to the maintainers only for them to fix my code for me.

File: quinn/src/connection.rs
Comment (FrankSpitulski) [reply quinn/src/connection.rs:463]: I don't think it should just be patched in locally. Hopefully I'm not the only one who will do development on CCAs. As for downcasting, that's not something I've needed to do. I just print out the debug representation.

File: quinn-proto/src/congestion/bbr.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion/bbr.rs:434]: I think BBR is pretty much done, but BBR2 is under development

File: quinn-proto/src/congestion/bbr_min_max.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion/bbr_min_max.rs]: done

File: quinn-proto/src/congestion/bbr.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion/bbr.rs:577]: yes. BBR ignores it. BBR2 does not.

File: quinn-proto/src/congestion.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion.rs:60]: yes, I will make the change

File: quinn-proto/src/congestion.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion.rs]: collapsed, but please check how it's invoked in quinn-proto/src/connection/mod.rs to make sure it still makes sense

File: quinn-proto/src/congestion/bbr.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion/bbr.rs:101]: fixed

File: quinn-proto/src/congestion/bbr.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion/bbr.rs]: fixed

File: quinn-proto/src/congestion/bbr.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion/bbr.rs]: changed description

File: quinn-proto/src/connection/mod.rs
Comment (FrankSpitulski) [reply quinn-proto/src/connection/mod.rs:135]: it's public so that I can do `self.0.lock("congestion").inner.path.congestion.clone_box()` in congestion_state()

Comment (FrankSpitulski) [reply quinn-proto/src/connection/mod.rs]: I've refactored it to make this not public.

File: quinn/src/connection.rs
Comment (Ralith) [reply quinn/src/connection.rs]: If the intended use is just for diagnostics during development, we should at least document that fact.

> I just print out the debug representation.

How does that work? `Controller` doesn't imply `Debug`.

File: quinn-proto/src/congestion/cubic.rs
Comment (Ralith) [inline quinn-proto/src/congestion/cubic.rs:167]: Wait, we also have an `#[allow]` on the actual `if` expression. Is it not effective? If so, we should remove it.

File: quinn-proto/src/congestion/cubic.rs
Comment (Ralith) [reply quinn-proto/src/congestion/cubic.rs:185]: I see there's actually already a report linked--not sure how I missed that.

File: quinn/src/connection.rs
Comment (FrankSpitulski) [reply quinn/src/connection.rs]: controller does imply debug since controller implements send + debug

File: quinn-proto/src/congestion.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion.rs]: combined into one

File: quinn/src/connection.rs
Comment (Ralith) [reply quinn/src/connection.rs]: Oh, right, that was added in this PR. Feels like a bit of a hack, but I guess it's harmless.

File: quinn-proto/src/congestion/cubic.rs
Comment (FrankSpitulski) [reply quinn-proto/src/congestion/cubic.rs:167]: I just rebased. Seems like @Matthias247 already covered it. I'll remove all related changes since it isn't related to BBR.

File: quinn/src/connection.rs
Comment (FrankSpitulski) [reply quinn/src/connection.rs]: it's definitely a hack- any better way to achieve printing this data out given a connection is  welcome

File: quinn/src/connection.rs
Comment (FrankSpitulski) [reply quinn/src/connection.rs]: continued discussion here https://github.com/quinn-rs/quinn/pull/1151#discussion_r667590209

**Comment by FrankSpitulski** [comment]
@Ralith is this still in the pipeline?

**Comment by Ralith** [comment]
Yeah, sorry, just on limited bandwidth at the moment.

**Comment by djc** [comment]
@FrankSpitulski this is ready for review, right? Mind marking it as such?

**Review by Ralith — DISMISSED** [review]
Thanks!

**Comment by FrankSpitulski** [comment]
had to rebase @Ralith

**Review by Ralith — APPROVED** [review]

**Pull request merged**
